### PR TITLE
 porting more fixes and improvement for aggregation

### LINKED ISSUE #614: Fixes and improve more for aggregation
needs to be back ported from community:

~~- [ ] https://github.com/ClickHouse/ClickHouse/pull/61257 (Fix possible incorrect result of aggregate function uniqExact #61257)~~
~~- [ ] https://github.com/ClickHouse/ClickHouse/pull/61171 (Remove a few templates from Aggregator.cpp #61171)~~
~~- [ ] https://github.com/ClickHouse/ClickHouse/pull/47399 (Remove distinct on top of MergingAggregatedStep #47399)~~
~~- [ ] https://github.com/ClickHouse/ClickHouse/pull/44541 (Followup #43761 #44541)~~
~~- [ ] https://github.com/ClickHouse/ClickHouse/pull/43761 (Refactor FunctionNode #43761)~~
- [x] https://github.com/ClickHouse/ClickHouse/pull/58542 (Revert "Merging #53757" #58542)
~~- [ ] https://github.com/ClickHouse/ClickHouse/pull/50531 (JIT infrastructure refactoring #50531)~~
~~- [ ] https://github.com/ClickHouse/ClickHouse/pull/43813 (Better handling of NULL in aggregate combinators, fix possible segfault/logical error #43813)~~
~~- [ ] https://github.com/ClickHouse/ClickHouse/pull/45067 (JIT compilation float to bool conversion fix #45067)~~
- [x] https://github.com/ClickHouse/ClickHouse/pull/59009 (uniqExact state parallel merging for distributed queries #59009)
- [x] https://github.com/ClickHouse/ClickHouse/pull/58745 (Fix double destroy call on exception throw in addBatchLookupTable8 #58745)
~~- [ ] https://github.com/ClickHouse/ClickHouse/pull/57872 (Better optimization of consecutive keys in aggregation #57872)~~
- [x] https://github.com/ClickHouse/ClickHouse/pull/43945 (Improve hash table preallocation optimisation)
- [x] https://github.com/ClickHouse/ClickHouse/pull/58440 (Enable JIT compilation for aggregation without key)
- [x] https://github.com/ClickHouse/ClickHouse/pull/57258 (Remove addBatchSinglePlaceFromInterval #57258)
- [x] https://github.com/ClickHouse/ClickHouse/pull/56120 (WindowTransform decrease amount of virtual function calls #56120)
~~- [ ] https://github.com/ClickHouse/ClickHouse/pull/57074 (Avoid excessive allocation in Arena #57074)~~
- [x] https://github.com/ClickHouse/ClickHouse/pull/53589 (Experiment Aggregator merge and destroy states in batch #53589)
~~- [ ] https://github.com/ClickHouse/ClickHouse/pull/53549 (Optimize group by constant keys #53549)~~
~~- [ ] https://github.com/ClickHouse/ClickHouse/pull/48607 (Add settings for real-time updates during query execution #48607)~~
- [x] https://github.com/ClickHouse/ClickHouse/pull/51566 (Fix memory consumption when max_block_size is huge in Aggregator + more memory checks #51566)
- [x] https://github.com/ClickHouse/ClickHouse/pull/50748 (Convert hashSets in parallel before merge)
- [x] https://github.com/ClickHouse/ClickHouse/pull/52439 (Fix possible double-free in Aggregator #52439)
- [x] https://github.com/ClickHouse/ClickHouse/pull/50291 (JIT aggregation nullable key fix #50291)
- [x] https://github.com/ClickHouse/ClickHouse/pull/49282 (Fix all key value is null and group use rollup return wrong answer #49282)
- [x] https://github.com/ClickHouse/ClickHouse/pull/45772 (Optimize one nullable key aggregate performance #45772)
- [x] https://github.com/ClickHouse/ClickHouse/pull/45375 (Window view functional test failed on platform s390x)
- [x] https://github.com/ClickHouse/ClickHouse/pull/40762 (Fix GROUPING function SQL compatibility #40762)
- [x] https://github.com/ClickHouse/ClickHouse/pull/39304 (Prefetching in aggregation #39304)
- [x] https://github.com/ClickHouse/ClickHouse/pull/39138 (Control block size in aggregator #39138)
- [x] https://github.com/ClickHouse/ClickHouse/pull/38295 (Quick fix for aggregation pipeline #38295)
- [x] https://github.com/ClickHouse/ClickHouse/pull/39908 (Improve 02354_distributed_with_external_aggregation_memory_usage #39908)
- [x] https://github.com/ClickHouse/ClickHouse/pull/39429 (Improve memory usage during memory efficient merging of aggregation results)
~~- [ ] https://github.com/ClickHouse/ClickHouse/pull/37959 (Preserve filling step plan position on sorting optimization #37959)~~
~~- [ ] https://github.com/ClickHouse/ClickHouse/pull/39123 (Pushdown filter to the right side of sorting join #39123)~~
~~- [ ] https://github.com/ClickHouse/ClickHouse/pull/37393 (Move updateInputStream to ITransformingStep #37393)~~
~~- [ ] https://github.com/ClickHouse/ClickHouse/pull/34632 (Optimized processing of ORDER BY in window functions #34632)~~
~~- [ ] https://github.com/ClickHouse/ClickHouse/pull/35623 (Functions calculation after sorting #35623)~~
~~- [ ] https://github.com/ClickHouse/ClickHouse/pull/34469 (Added JIT compilation of SortDescription #34469)~~
- [x] https://github.com/ClickHouse/ClickHouse/pull/37163 (Add GROUPING function #37163)
- [x] https://github.com/ClickHouse/ClickHouse/pull/37469 (Implement in order aggregation (optimize_aggregation_in_order) for projections for tables with fully materialized projections #37469)
- [x] https://github.com/ClickHouse/ClickHouse/pull/34305 (Fix various issues when projection is enabled by default #34305)
- [x] https://github.com/ClickHouse/ClickHouse/pull/37093 (Fix GROUP BY AggregateFunction #37093)

### PR DESCRIPTION
**Pull request by yl-lisen**
PR checklist:
- Did you run ClangFormat ?
- Did you separate headers to a different section in existing community code base ?
- Did you surround `proton: starts/ends` for new code in existing community code base ?

Please write user-readable short description of the changes:
This close #614

### CODE REVIEW

File: src/AggregateFunctions/AggregateFunctionUniq.h [hunk 1]
```diff
@@ -96,18 +104,33 @@ struct AggregateFunctionUniqHLL12Data<UUID, false>
...
 
     constexpr static bool is_able_to_parallelize_merge = false;
+    constexpr static bool is_parallelize_merge_prepare_needed = false;
     constexpr static bool is_variadic = false;
 
     static String getName() { return "unique_hll12"; }
 };
 
+template <>
+struct AggregateFunctionUniqHLL12Data<IPv6, false>
```
Comment (yokofly) [inline src/AggregateFunctions/AggregateFunctionUniq.h]: ipv6 change introduce in this commit https://github.com/timeplus-io/proton/commit/7411080d85ffc0a2981b5d5f42e2a52d3a2ffaf6#diff-28d726caaebff128ca648d66098ce3f23cda1e924336a302934a4829c6680608R114

but actually, this `ipv6` specialization is part of another big pr 
https://github.com/ClickHouse/ClickHouse/pull/50551/files#diff-28d726caaebff128ca648d66098ce3f23cda1e924336a302934a4829c6680608R105

I recommend either porting this full pr, or we remove here?

File: tests/queries/0_stateless/01710_projection_aggregation_in_order.sql [hunk 2]
```diff
@@ -31,8 +31,11 @@ SET allow_experimental_projection_optimization=1, optimize_aggregation_in_order=
 WITH toStartOfHour(ts) AS a SELECT sum(value) v FROM normal WHERE ts > '2021-12-06 22:00:00' GROUP BY a ORDER BY v LIMIT 5;
 WITH toStartOfHour(ts) AS a SELECT sum(value) v FROM normal WHERE ts > '2021-12-06 22:00:00' GROUP BY toStartOfHour(ts), a ORDER BY v LIMIT 5;
 
+<<<<<<< HEAD
```
Comment (yokofly) [inline tests/queries/0_stateless/01710_projection_aggregation_in_order.sql]: clean this. if we have no time to run this stateless let's post it in tests/queries_not_supported folder.
proton-client supports `sql` file input.
it is better to manually run the statless tests to make sure we do not break the compatibility in later code changes.

File: tests/queries/0_stateless/01710_projection_aggregation_in_order.sql [hunk 3]
```diff
@@ -60,7 +63,12 @@ FROM numbers(100000);
 
 SET allow_experimental_projection_optimization=1, optimize_aggregation_in_order=1, force_optimize_projection = 1;
 
+<<<<<<< HEAD
 WITH toStartOfHour(ts) AS a SELECT sum(value) v FROM agg WHERE ts > '2021-12-06 22:00:00' GROUP BY a ORDER BY v LIMIT 5;
 WITH toStartOfHour(ts) AS a SELECT sum(value) v FROM agg WHERE ts > '2021-12-06 22:00:00' GROUP BY toStartOfHour(ts), a ORDER BY v LIMIT 5;
 
 DROP TABLE agg;
+=======
```
Comment (yokofly) [inline tests/queries/0_stateless/01710_projection_aggregation_in_order.sql]: same, remove here plz

**Review by yokofly — COMMENTED** [review]
shall we consider this pr? https://github.com/ClickHouse/ClickHouse/pull/60138/files
issue wanna solve:
> We should have some sanity check around max_block_size since it has a direct relationship with memory usage.

File: src/AggregateFunctions/AggregateFunctionUniq.h (hunk 1 above)
Comment (yl-lisen) [reply src/AggregateFunctions/AggregateFunctionUniq.h]: will remove it

**Comment by yl-lisen** [comment]
> shall we consider this pr? https://github.com/ClickHouse/ClickHouse/pull/60138/files issue wanna solve:
> 
> > We should have some sanity check around max_block_size since it has a direct relationship with memory usage.

This checking logic needs to rely on some code base. It is not recommended to be included in this PR, we can do it in a separate PR.

**Review by yokofly — APPROVED** [review]

**Pull request merged**